### ISSUE TITLE
Bond iface does not work when active slave is teared down #12

### ISSUE THREAD

**Issue by alosadagrande** [intel/bond-cni#14]
Update: similar issue to #11 - in my case I see this behaviour when I tear down the net1 which is the active slave interface.

Hi,

I have created a bond interface composed by two sr-iov slaves. I can see that the bond interface is successfully connected to the physical network and the load balance mode set to active-passive is working as expected.

These are the NAD and the pod definitions:

```sh
oc get -o yaml net-attach-def bond-dhcp
apiVersion: k8s.cni.cncf.io/v1
kind: NetworkAttachmentDefinition
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"k8s.cni.cncf.io/v1","kind":"NetworkAttachmentDefinition","metadata":{"annotations":{},"name":"bond-dhcp","namespace":"sriov-test-bond"},"spec":{"config":"{ \"type\": \"bond\", \"cniVersion\": \"0.3.1\", \"name\": \"bond-net1\", \"ifname\": \"net3\", \"spoofChk\": \"off\", \"mode\": \"active-backup\", \"linksInContainer\": true, \"miimon\": \"100\", \"links\": [ {\"name\": \"net1\"}, {\"name\": \"net2\"} ], \"ipam\": { \"type\":\"dhcp\"\n} }"}}
  creationTimestamp: "2020-05-04T10:08:13Z"
  generation: 2
  name: bond-dhcp
  namespace: sriov-test-bond
  resourceVersion: "4331357"
  selfLink: /apis/k8s.cni.cncf.io/v1/namespaces/sriov-test-bond/network-attachment-definitions/bond-dhcp
  uid: 7fdc413f-16bb-4044-a8ef-666fa6717922
spec:
  config: |-
    { "type": "bond", "cniVersion": "0.3.1", "name": "bond-net1", "ifname": "net3", "spoofChk": "off", "mode": "active-backup", "linksInContainer": true, "miimon": "100", "links": [ {"name": "net1"}, {"name": "net2"} ], "ipam": { "type":"dhcp"
    } }

```
apiVersion: v1
kind: Pod
metadata:
  name: bond-pod-dhcp
  namespace: sriov-test-bond
  annotations:
    k8s.v1.cni.cncf.io/networks: slaves, slaves, bond-dhcp
spec:
  containers:
  - name: podexample
    securityContext:
      privileged: true
    image: quay.io/schseba/utility-container:latest
    command: ["/bin/bash", "-c", "sleep INF"]
```

The pod is able to get an IP from the DHCP server:

```sh
oc rsh bond-pod-dhcp ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
3: eth0@if9459: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1400 qdisc noqueue state UP group default 
    link/ether 16:f4:4c:87:00:06 brd ff:ff:ff:ff:ff:ff link-netnsid 0
    inet 10.135.0.5/23 brd 10.135.1.255 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::44a8:8dff:fe66:2437/64 scope link 
       valid_lft forever preferred_lft forever
4: net3: <BROADCAST,MULTICAST,MASTER,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether d6:51:55:dd:09:f4 brd ff:ff:ff:ff:ff:ff
    inet 172.22.0.63/24 scope global net3
       valid_lft forever preferred_lft forever
    inet6 fe80::d451:55ff:fedd:9f4/64 scope link 
       valid_lft forever preferred_lft forever
20: net1: <BROADCAST,MULTICAST,SLAVE,UP,LOWER_UP> mtu 1500 qdisc mq master net3 state UP group default qlen 1000
    link/ether d6:51:55:dd:09:f4 brd ff:ff:ff:ff:ff:ff
21: net2: <BROADCAST,MULTICAST,SLAVE,UP,LOWER_UP> mtu 1500 qdisc mq master net3 state UP group default qlen 1000
    link/ether d6:51:55:dd:09:f4 brd ff:ff:ff:ff:ff:ff
```

And able to reach the network:

```sh
oc rsh bond-pod-dhcp ping 172.22.0.253
PING 172.22.0.253 (172.22.0.253) 56(84) bytes of data.
64 bytes from 172.22.0.253: icmp_seq=1 ttl=64 time=0.205 ms
64 bytes from 172.22.0.253: icmp_seq=2 ttl=64 time=0.256 ms
```
The bond configuration is the expected one as well:

```sh
sh-4.2# cat /proc/net/bonding/net3 
Ethernet Channel Bonding Driver: v3.7.1 (April 27, 2011)

Bonding Mode: fault-tolerance (active-backup)
Primary Slave: None
Currently Active Slave: net1
MII Status: up
MII Polling Interval (ms): 100
Up Delay (ms): 0
Down Delay (ms): 0

Slave Interface: net1
MII Status: up
Speed: 25000 Mbps
Duplex: full
Link Failure Count: 0
Permanent HW addr: d6:51:55:dd:09:f4
Slave queue ID: 0

Slave Interface: net2
MII Status: up
Speed: 25000 Mbps
Duplex: full
Link Failure Count: 0
Permanent HW addr: 62:17:0c:6e:b6:f0
Slave queue ID: 0
```

If I tear down the net1 interface the net2 becomes the active one, but I am not able to connect to the network:

```sh
sh-4.2# ip link set net1 down

sh-4.2# ping 172.22.0.253
PING 172.22.0.253 (172.22.0.253) 56(84) bytes of data.
From 172.22.0.63 icmp_seq=10 Destination Host Unreachable
From 172.22.0.63 icmp_seq=11 Destination Host Unreachable
From 172.22.0.63 icmp_seq=12 Destination Host Unreachable

sh-4.2# cat /proc/net/bonding/net3 
Ethernet Channel Bonding Driver: v3.7.1 (April 27, 2011)

Bonding Mode: fault-tolerance (active-backup)
Primary Slave: None
Currently Active Slave: net2
MII Status: up
MII Polling Interval (ms): 100
Up Delay (ms): 0
Down Delay (ms): 0

Slave Interface: net1
MII Status: down
Speed: 25000 Mbps
Duplex: full
Link Failure Count: 1
Permanent HW addr: d6:51:55:dd:09:f4
Slave queue ID: 0

Slave Interface: net2
MII Status: up
Speed: 25000 Mbps
Duplex: full
Link Failure Count: 0
Permanent HW addr: 62:17:0c:6e:b6:f0
Slave queue ID: 0
```
dmesg output:

```sh
[518128.738347] net3: link status definitely down for interface net1, disabling it
[518128.745685] net3: making interface net2 the new active one
[518147.220852] device f07c8aba10ef6a0 left promiscuous mode
[518155.644880] SELinux: mount invalid.  Same superblock, different security settings for (dev mqueue, type mqueue)
[518155.648273] SELinux: mount invalid.  Same superblock, different security settings for (dev mqueue, type mqueue)
[518155.911369] IPv6: ADDRCONF(NETDEV_UP): eth0: link is not ready
[518155.938309] IPv6: ADDRCONF(NETDEV_UP): veth73f8e2a0: link is not ready
[518155.945130] IPv6: ADDRCONF(NETDEV_CHANGE): veth73f8e2a0: link becomes ready
[518155.947316] SELinux: mount invalid.  Same superblock, different security settings for (dev mqueue, type mqueue)
[518155.952284] IPv6: ADDRCONF(NETDEV_CHANGE): eth0: link becomes ready
[518685.844857] 4e518eaf8d029fd: renamed from vethfbeea613
[518685.895350] device 4e518eaf8d029fd entered promiscuous mode
[518687.915377] SELinux: mount invalid.  Same superblock, different security settings for (dev mqueue, type mqueue)
[518695.548931] SELinux: mount invalid.  Same superblock, different security settings for (dev mqueue, type mqueue)
[518695.822684] SELinux: mount invalid.  Same superblock, different security settings for (dev mqueue, type mqueue)
[518711.233240] SELinux: mount invalid.  Same superblock, different security settings for (dev mqueue, type mqueue)
[518735.567543] SELinux: mount invalid.  Same superblock, different security settings for (dev mqueue, type mqueue)
[518735.851176] SELinux: mount invalid.  Same superblock, different security settings for (dev mqueue, type mqueue)
[518807.358873] device 4e518eaf8d029fd left promiscuous mode
```

Notes:

- If I tear down the net2 interface and activate net1 again, so the net1 becomes the active one in the bonding, I am able to connect to the network again.

- The spoofChk is set to off as I can see from the node vf 0 and 3:

2: eno1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP mode DEFAULT group default qlen 1000
    link/ether 98:03:9b:61:ba:d0 brd ff:ff:ff:ff:ff:ff
    vf 0 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust on, query_rss off 
    vf 1 MAC 00:00:00:00:00:00, spoof checking off, link-state auto, trust on, query_rss off
    vf 2 MAC 00:00:00:00:00:00, spoof checking off, link-state auto, trust on, query_rss off
    vf 3 MAC 00:00:00:00:00:00, spoof checking on, link-state enable, trust on, query_rss off
    vf 4 MAC 00:00:00:00:00:00, spoof checking on, link-state auto, trust on, query_rss off

-  dmesg does not say anything special apart from showing that the active iface now is net2.

Questions:

- Looks like to be something related to spoofing, probably the MAC address of the bond interface must change if the net2 is the active one? or in the other way around? I do not think there is something with port-security at switch level...

**Comment by killianmuldoon** [comment]
I've added failOverMac as a configuraiton option which seems to solve this issue - @alosadagrande could you try pulling the PR and let me know if it works for your issue?

I've got a bond network attachment definition with:

```
apiVersion: "k8s.cni.cncf.io/v1"
kind: NetworkAttachmentDefinition
metadata:
  name: bond-net
spec:
  config: '{
  "type": "bond",
  "cniVersion": "0.3.1",
  "name": "bond-net",
  "ifname": "bond0",
  "mode": "active-backup",
  "failOverMac": 1,
  "linksInContainer": true,
  "miimon": "100",
  "links": [
     {"name": "net1"},
     {"name": "net2"}
  ],
  "ipam": {
    "type": "host-local",
    "subnet": "10.56.217.0/24",
    "routes": [{
      "dst": "0.0.0.0/0"
    }],
    "gateway": "10.56.217.1"
  }
}'
```

Note this is using two kernel driver sriov functions, with net1 and net2 found in the container. 

When I run using this configuration on my machine (in an unprivileged pod) I get:

**Initial bonding configuration**
```
#kubectl exec -it test-pod -- cat /proc/net/bonding/bond0
Ethernet Channel Bonding Driver: v3.7.1 (April 27, 2011)

Bonding Mode: fault-tolerance (active-backup) (fail_over_mac active)
Primary Slave: None
Currently Active Slave: net1
MII Status: up
MII Polling Interval (ms): 100
Up Delay (ms): 0
Down Delay (ms): 0

Slave Interface: net1
MII Status: up
Speed: 40000 Mbps
Duplex: full
Link Failure Count: 0
Permanent HW addr: 6a:95:79:e3:64:68
Slave queue ID: 0

Slave Interface: net2
MII Status: up
Speed: 40000 Mbps
Duplex: full
Link Failure Count: 0
Permanent HW addr: 7e:ca:9c:ed:23:b5
Slave queue ID: 0

```
**Initial ping (to outside sriov net container on same host)**
```
#kubectl exec -it test-pod -- ping 10.244.0.91 
PING 10.244.0.91 (10.244.0.91): 56 data bytes
64 bytes from 10.244.0.91: seq=0 ttl=64 time=0.139 ms
```
**Turning off net1** ( using netns as we don't have priviledge in pod):

```
#nsenter -t ${PID of container from docker inspect) -n cat /proc/net/bonding/bond0

```
**New bond config**
```
#nsenter -t 68543 -n cat /proc/net/bonding/bond0
Ethernet Channel Bonding Driver: v3.7.1 (April 27, 2011)

Bonding Mode: fault-tolerance (active-backup) (fail_over_mac active)
Primary Slave: None
Currently Active Slave: net2
MII Status: up
MII Polling Interval (ms): 100
Up Delay (ms): 0
Down Delay (ms): 0

Slave Interface: net1
MII Status: down
Speed: 40000 Mbps
Duplex: full
Link Failure Count: 1
Permanent HW addr: 6a:95:79:e3:64:68
Slave queue ID: 0

Slave Interface: net2
MII Status: up
Speed: 40000 Mbps
Duplex: full
Link Failure Count: 0
Permanent HW addr: 7e:ca:9c:ed:23:b5
Slave queue ID: 0
```

**Final ping test**
```
#kubectl exec -it test-pod -- ping 10.244.0.91
PING 10.244.0.91 (10.244.0.91): 56 data bytes
64 bytes from 10.244.0.91: seq=0 ttl=64 time=0.184 ms
64 bytes from 10.244.0.91: seq=1 ttl=64 time=0.172 ms
^C
--- 10.244.0.91 ping statistics ---
2 packets transmitted, 2 packets received, 0% packet loss
round-trip min/avg/max = 0.172/0.178/0.184 ms
```

**Comment by killianmuldoon** [comment]
Closing as issue appears solved in PR #11
